### ISSUE TITLE
[BUG] thread issues with evaluation

### ISSUE THREAD

**Issue by acsweet** [ml-explore/mlx#2067]
**Describe the bug**
There seem to be some thread data access issues introduced in the newest version of mlx.

**To Reproduce**
This doesn't produce the exact same error I was experiencing, but it does produce a similar error (I think).
```python
import mlx.core as mx
import mlx.nn as nn
import mlx.optimizers as optim
import numpy as np
import threading
import time
import traceback
import faulthandler

faulthandler.enable()

class SimpleModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.dense1 = nn.Linear(10, 32)
        self.dense2 = nn.Linear(32, 16)
        self.dense3 = nn.Linear(16, 1)
        
    def __call__(self, x):
        x = mx.maximum(0, self.dense1(x))
        x = mx.maximum(0, self.dense2(x))
        return self.dense3(x)

def loss_fn(model, x, y):
    pred = model(x)
    return mx.mean((pred - y) ** 2)

def train_and_convert(thread_id):
    try:
        print(f"Thread {thread_id} starting")
        
        model = SimpleModel()
        optimizer = optim.Adam(learning_rate=0.001)

        x = mx.random.normal((128, 10))
        y = mx.random.normal((128, 1))
        
        for i in range(10):
            print(f"Thread {thread_id}, Iteration {i} starting")
            loss, grads = mx.value_and_grad(loss_fn)(model, x, y)
            
            print(f"Thread {thread_id}, Iteration {i} - evaluating gradients")
            for g in grads.values():
                mx.eval(g)
            
            print(f"Thread {thread_id}, Iteration {i} - updating model")
            optimizer.update(model, grads)
            
            print(f"Thread {thread_id}, Iteration {i} - evaluating loss")
            mx.eval(loss)
            
            print(f"Thread {thread_id}, Iteration {i} - converting loss to numpy")
            # Convert loss to numpy
            np_loss = np.array(loss)
            
            print(f"Thread {thread_id}, Iteration {i} - converting gradients to numpy")
            for name, g in list(grads.items())[:2]:
                print(f"Thread {thread_id}, Iteration {i} - converting gradient {name}")
                np_grad = np.array(g)
                print(f"Thread {thread_id}, Iteration {i}, Grad {name} shape: {np_grad.shape}")
            
            print(f"Thread {thread_id}, Iteration {i}, Loss: {np_loss}")
            
            # small delay to increase chances of thread overlap
            time.sleep(0.01)
    except Exception as e:
        print(f"Thread {thread_id} failed with exception: {e}")
        print(traceback.format_exc())

threads = []
for i in range(3): # Try with more threads
    t = threading.Thread(target=train_and_convert, args=(i,))
    t.daemon = False
    threads.append(t)
    t.start()
    # Small delay between thread starts
    time.sleep(0.05)

try:
    for t in threads:
        t.join()
    print("All threads completed successfully")
except KeyboardInterrupt:
    print("Interrupted by user")
except Exception as e:
    print(f"Exception in main thread: {e}")
    print(traceback.format_exc())
```

**Expected behavior**
When I run this with mlx `0.23.2` it executes fine, but with `0.24.0` and up it either gives a segmentation fault or python fatal error like:
```
python(15651,0x171ba3000) malloc: Double free of object 0x14ea0dbb0
python(15651,0x171ba3000) malloc: *** set a breakpoint in malloc_error_break to debug
Fatal Python error: Aborted
```
I'd expect it to execute with no issues, or if this is expected/unavoidable behavior with the updates from version `0.24.0` onwards.

**Desktop (please complete the following information):**
 - OS Version: MacOS 15.2
 - Version: 0.23.2 and 0.24.0 up

**Additional context**
This is related to the mlx backend for Keras (https://github.com/keras-team/keras/issues/19571), and I initially ran across this error with a simple `model.fit()`, it would occur when the progress bar was updating (which involves a cast to a numpy array to display the loss). In that case it also executed to completion with `0.23.2` and returned for `0.24.2` failed with a segmentation fault or an error like this:
```
-[AGXG16XFamilyCommandBuffer tryCoalescingPreviousComputeCommandEncoderWithConfig:nextEncoderClass:]:1091: failed assertion `A command encoder is already encoding to this command buffer'
Fatal Python error: Aborted
```
While I was trying to replicate that last error I was able to get the error above. I can keep trying to replicate the last error with a simple code block if that's helpful, but hopefully the above is enough to identify what's happening.

**Comment by acsweet** [comment]
I think I'll see if I can work on a fix for this.

**Comment by awni** [comment]
Mlx is in general not thread safe. There isn’t an easy fix for this. You can play around with it and see what you think. But in general we’ll probably need to make some design decisions around if and how to support this kind of use case. We can use this issue to discuss further.

**Comment by acsweet** [comment]
If I have an idea should I submit a PR? And we can discuss what you think?

**Comment by awni** [comment]
Yea that's totally fine

**Comment by Armanoide** [comment]
Maybe it's a bit late to reply to your issue, but I had a similar error in Rust.

The mlx array is not thread-safe.
See: [Doc Swift](https://swiftpackageindex.com/ml-explore/mlx-swift/0.21.3/documentation/mlxrandom/randomstate)

To workaround your issue, you need to ensure that only one operation is sent to the GPU at a time.
The best way to achieve this is by using a mutex to make other threads wait until the current one finishes.

Here’s an example using Python:

```
from threading import Lock

mutex = Lock()
.....
class SimpleModel(nn.Module):
def train_and_convert(thread_id):
         mutex.acquire()
         mx.eval(loss)
         mutex.release()
......
```
same for the other `eval` i did't test but should work

**Comment by ngoldbaum** [comment]
When I trigger this, I see the following traceback inside of lldb:

<details>

```
-[AGXG15XFamilyCommandBuffer tryCoalescingPreviousComputeCommandEncoderWithConfig:nextEncoderClass:]:1091: failed assertion `A command encoder is already encoding to this command buffer'
Process 32498 stopped
* thread #4, name = 'ThreadPoolExecutor-0_0', stop reason = hit program assert
    frame #4: 0x00000001a7a97cb0 Metal`MTLReportFailure.cold.1 + 48
Metal`bool MTLGetEnvCase<MTLErrorModeType>(char const*, MTLErrorModeType&, std::__1::vector<std::__1::pair<char const*, MTLErrorModeType>, std::__1::allocator<std::__1::pair<char const*, MTLErrorModeType>>> const&) (.cold.1):
->  0x1a7a97cb0 <+0>:  pacibsp
    0x1a7a97cb4 <+4>:  sub    sp, sp, #0x40
    0x1a7a97cb8 <+8>:  stp    x22, x21, [sp, #0x10]
    0x1a7a97cbc <+12>: stp    x20, x19, [sp, #0x20]
Target 0: (python) stopped.
(lldb) bt
* thread #4, name = 'ThreadPoolExecutor-0_0', stop reason = hit program assert
    frame #0: 0x000000019c4ba388 libsystem_kernel.dylib`__pthread_kill + 8
    frame #1: 0x000000019c4f388c libsystem_pthread.dylib`pthread_kill + 296
    frame #2: 0x000000019c3fca3c libsystem_c.dylib`abort + 124
    frame #3: 0x000000019c3fbc70 libsystem_c.dylib`__assert_rtn + 284
  * frame #4: 0x00000001a7a97cb0 Metal`MTLReportFailure.cold.1 + 48
    frame #5: 0x00000001a7a6cd70 Metal`MTLReportFailure + 576
    frame #6: 0x0000000101007ca8 AGXMetalG15X_M1`-[AGXG15XFamilyCommandBuffer tryCoalescingPreviousComputeCommandEncoderWithConfig:nextEncoderClass:] + 484
    frame #7: 0x0000000101007a5c AGXMetalG15X_M1`-[AGXG15XFamilyCommandBuffer computeCommandEncoderWithConfig:] + 84
    frame #8: 0x0000000101007988 AGXMetalG15X_M1`-[AGXG15XFamilyCommandBuffer computeCommandEncoderWithDispatchType:] + 140
    frame #9: 0x00000001068deec4 libmlx.dylib`MTL::ComputeCommandEncoder* NS::Object::sendMessage<MTL::ComputeCommandEncoder*, MTL::DispatchType>(pObj=0x000000011b7085c0, selector="computeCommandEncoderWithDispatchType:", args=DispatchTypeConcurrent) at NSObject.hpp:216:16
    frame #10: 0x00000001068deea8 libmlx.dylib`MTL::CommandBuffer::computeCommandEncoder(this=0x000000011b7085c0, dispatchType=DispatchTypeConcurrent) at MTLCommandBuffer.hpp:333:12 [inlined]
    frame #11: 0x00000001068dee90 libmlx.dylib`mlx::core::metal::CommandEncoder::CommandEncoder(this=0x000060000192c1c0, stream=0x0000600001b200e8) at device.cpp:237:26
    frame #12: 0x00000001068df058 libmlx.dylib`mlx::core::metal::CommandEncoder::CommandEncoder(this=0x000060000192c1c0, stream=0x0000600001b200e8) at device.cpp:236:72
    frame #13: 0x00000001068e26c8 libmlx.dylib`std::__1::unique_ptr<mlx::core::metal::CommandEncoder, std::__1::default_delete<mlx::core::metal::CommandEncoder>> std::__1::make_unique[abi:ne200100]<mlx::core::metal::CommandEncoder, mlx::core::metal::DeviceStream&, 0>(__args=0x0000600001b200e8) at unique_ptr.h:767:30
    frame #14: 0x00000001068e25d4 libmlx.dylib`mlx::core::metal::Device::get_command_encoder(this=0x0000000106b98a00, index=0) at device.cpp:496:22
    frame #15: 0x000000010689f2b4 libmlx.dylib`mlx::core::binary_op_gpu_inplace(inputs=size=2, outputs=size=1, op="Add", s=0x0000600002ef0dd0) at binary.cpp:109:29
    frame #16: 0x00000001068a0230 libmlx.dylib`mlx::core::binary_op_gpu_inplace(inputs=size=2, out=0x000060000223c4e0, op="Add", s=0x0000600002ef0dd0) at binary.cpp:193:3
    frame #17: 0x00000001068a03e4 libmlx.dylib`mlx::core::binary_op_gpu(inputs=size=2, out=0x000060000223c4e0, op="Add", s=0x0000600002ef0dd0) at binary.cpp:206:3
    frame #18: 0x00000001068a0484 libmlx.dylib`mlx::core::binary_op_gpu(inputs=size=2, out=0x000060000223c4e0, op="Add") at binary.cpp:214:3
    frame #19: 0x00000001068a04d8 libmlx.dylib`mlx::core::Add::eval_gpu(this=0x0000600002ef0dc8, inputs=size=2, out=0x000060000223c4e0) at binary.cpp:217:1
    frame #20: 0x0000000105a54280 libmlx.dylib`mlx::core::UnaryPrimitive::eval_gpu(this=0x0000600002ef0dc8, inputs=size=2, outputs=size=1) at primitives.h:144:5
    frame #21: 0x000000010691839c libmlx.dylib`mlx::core::gpu::eval(arr=0x0000000170f1c158) at eval.cpp:48:21
    frame #22: 0x0000000105caf368 libmlx.dylib`mlx::core::eval_impl(outputs=size=0, async=false) at transforms.cpp:237:7
    frame #23: 0x0000000105cb0fe4 libmlx.dylib`mlx::core::eval(outputs=size=0) at transforms.cpp:324:3
    frame #24: 0x00000001059d60cc libmlx.dylib`mlx::core::array::eval(this=0x0000037c160c0728) at array.cpp:153:5
    frame #25: 0x00000001017d4af0 core.cpython-314t-darwin.so`to_scalar(a=0x0000037c160c0728) at convert.cpp:178:7
    frame #26: 0x00000001017a29e0 core.cpython-314t-darwin.so`_object* nanobind::detail::func_create<false, true, nanobind::object (*&)(mlx::core::array&), nanobind::object, mlx::core::array&, 0ul, nanobind::scope, nanobind::name, nanobind::is_method, nanobind::sig, char [122]>(nanobind::object (*&)(mlx::core::array&), nanobind::object (*)(mlx::core::array&), std::__1::integer_sequence<unsigned long, 0ul>, nanobind::scope const&, nanobind::name const&, nanobind::is_method const&, nanobind::sig const&, char const (&) [122])::'lambda'(void*, _object**, unsigned char*, nanobind::rv_policy, nanobind::detail::cleanup_list*)::operator()(this=0x0000000170f1c8cf, p=0x0000037c12f2f2b8, args=0x0000000170f1d150, args_flags="\U00000001\U00000001\U00000001\U00000001\U00000001\U00000001\U00000001\U00000001l", policy=automatic, cleanup=0x0000000170f1caa0) const at nb_func.h:275:24 [inlined]
    frame #27: 0x00000001017a290c core.cpython-314t-darwin.so`_object* nanobind::detail::func_create<false, true, nanobind::object (*&)(mlx::core::array&), nanobind::object, mlx::core::array&, 0ul, nanobind::scope, nanobind::name, nanobind::is_method, nanobind::sig, char [122]>(nanobind::object (*&)(mlx::core::array&), nanobind::object (*)(mlx::core::array&), std::__1::integer_sequence<unsigned long, 0ul>, nanobind::scope const&, nanobind::name const&, nanobind::is_method const&, nanobind::sig const&, char const (&) [122])::'lambda'(void*, _object**, unsigned char*, nanobind::rv_policy, nanobind::detail::cleanup_list*)::__invoke(p=0x0000037c12f2f2b8, args=0x0000000170f1d150, args_flags="\U00000001\U00000001\U00000001\U00000001\U00000001\U00000001\U00000001\U00000001l", policy=automatic, cleanup=0x0000000170f1caa0) at nb_func.h:219:14
    frame #28: 0x0000000101a30f50 core.cpython-314t-darwin.so`nanobind::detail::nb_func_vectorcall_simple(self=0x0000037c12f2f280, args_in=0x0000000170f1d150, nargsf=9223372036854775809, kwargs_in=0x0000000000000000) at nb_func.cpp:914:26
    frame #29: 0x00000001007a5f98 libpython3.14t.dylib`PyObject_Vectorcall + 88
    frame #30: 0x00000001008f82ec libpython3.14t.dylib`_PyEval_EvalFrameDefault + 9060
    frame #31: 0x00000001008f5ce0 libpython3.14t.dylib`_PyEval_Vector + 716
    frame #32: 0x00000001007a941c libpython3.14t.dylib`method_vectorcall + 444
    frame #33: 0x00000001008f9dc8 libpython3.14t.dylib`_PyEval_EvalFrameDefault + 15936
    frame #34: 0x00000001008f5ce0 libpython3.14t.dylib`_PyEval_Vector + 716
    frame #35: 0x00000001007a5658 libpython3.14t.dylib`_PyObject_VectorcallDictTstate + 204
    frame #36: 0x00000001007a688c libpython3.14t.dylib`_PyObject_Call_Prepend + 152
    frame #37: 0x0000000100858284 libpython3.14t.dylib`call_method + 124
    frame #38: 0x00000001007a5804 libpython3.14t.dylib`_PyObject_MakeTpCall + 328
    frame #39: 0x00000001008f82ec libpython3.14t.dylib`_PyEval_EvalFrameDefault + 9060
    frame #40: 0x00000001008f5ce0 libpython3.14t.dylib`_PyEval_Vector + 716
    frame #41: 0x00000001007a939c libpython3.14t.dylib`method_vectorcall + 316
    frame #42: 0x0000000100928a5c libpython3.14t.dylib`context_run + 136
    frame #43: 0x00000001007a5f98 libpython3.14t.dylib`PyObject_Vectorcall + 88
    frame #44: 0x00000001008f82ec libpython3.14t.dylib`_PyEval_EvalFrameDefault + 9060
    frame #45: 0x00000001008f5ce0 libpython3.14t.dylib`_PyEval_Vector + 716
    frame #46: 0x00000001007a939c libpython3.14t.dylib`method_vectorcall + 316
    frame #47: 0x0000000100a17868 libpython3.14t.dylib`thread_run + 128
    frame #48: 0x0000000100999d04 libpython3.14t.dylib`pythread_wrapper + 28
    frame #49: 0x000000019c4f3c0c libsystem_pthread.dylib`_pthread_start + 136
```

</details>

`get_command_encoder` looks a little suspicious to me, and indeed:

https://github.com/ml-explore/mlx/blob/012fb220a19035c2ca28ddcf8b2d204eaaf7e761/mlx/backend/cpu/encoder.cpp#L7-L14

the static unordered_map is not initialized or accessed in a thread-safe manner. It probably needs a lock?

I think ideally Python users would see a Python exception telling them that MLX isn't thread-safe and they need to do python-level synchronization. I'm not sure what the goal is for C++ users, although I gather from those Swift docs there probably isn't any intention to make the Python API thread-safe.

Unfortunately I don't think it's tenable to do this at the Python level so it's probably necessary to fix some thread safety issues in the C++ library to at least make it possible to produce nicer errors and avoid the asserts in the Metal implementation we're currently hitting doing pretty much anything nontrivial with threads and MLX involved.

**Comment by zcbenz** [comment]
Closing since MLX has thread safety support in 0.31.2 now, if you encountered problems please open new issues.
